feat(readers): add DataSource implementations

yeewen1018 opened the pull request:
This PR implements the first batch of `DataSource` readers
* **TxtSource** – streams line by line from `.txt` files, skips blank lines.
* **JsonlSource** – parses JSON-per-line into `serde_json::Value` or a user-supplied type.
* **ImageDirSource** – stream image files from a directory
* **ParquetSource** – streams Arrow `RecordBatch`es .
* **SafetensorsSource** – loads tensors from a `.safetensors` file

### Tests
- Added unit tests for each `DataSource` implemented. 

### Dependencies
* **serde_json**, **serde**, **safetensors**, **parquet**, **arrow**, **bytemuck**

### Proposed next steps
1. Work on `Transform` -> `Sampler` -> `DataLoader` 
2. After that, extend on `DataSources` to support more sources (HTTP, SQLite, HDF5, S3, tar.gz, video frames).  

_This PR depends on #22 (dataset-trait)._

yeewen1018 committed:
feat: add Dataset trait and make Sample struct Send+Sync

yeewen1018 committed:
feat(dataset): implement InMemoryDataset

yeewen1018 committed:
test(dataset): add unit tests for InMemoryDataset

yeewen1018 committed:
test(dataset): add pipeline integration tests

Add end-to-end validation for:
- InMemoryDataset sample loading/iteration
- Stack collation overflow
- Padding collation with variable-length samples
- Device transfer roundtrips (CPU<->GPU)

yeewen1018 committed:
feat(dataset): refine Dataset trait docs and change get() return type
- Update documentation on the `Dataset` trait for `get`/`len` to clarify the expected behaviour of InMemoryDataset and IterableDataset.
- Change `get` function from `Option<Result<&Sample>>` to `Result<Option<&Sample>>` for better error handling.
- Update all affected tests to match the new `get` function.

yeewen1018 committed:
feat(dataset): add opt-in boxed iterator via `boxed-iter` feature.

yeewen1018 committed:
bench: add InMemoryDataset iteration benchmarks for static vs boxed iteration.

yeewen1018 committed:
refactor(dataset): remove `len()`, `get()` and `is_empty()` from Dataset trait

- Add inherent `get(&self, idx) -> Option<&Sample>`, `len(&self) -> usize`, and `is_empty(&self) -> bool` to `InMemoryDataset`
- Update `len()` to return `usize` instead of `Option(usize)`
- Update tests to return `usize` for `len()` assertions

yeewen1018 committed:
feat(dataset): Add IterableDataset for streaming support

- Add `DataSource` enum (to support various types of data sources)
- Implement IterableDataset with dummy per-source `stream_samples()` method
- Include basic test for `IterableDatset::iter()`

yeewen1018 committed:
refactor: remove boxed-iter feature flag
- Make boxed iterator always available

yeewen1018 committed:
refactor(benches): Clarify benchmark variable names
- b -> bench (benchmark harness)
- r -> sample_result (iterator item)
- t -> processed_tensor (transformed tensor)
- ds -> dataset
- &_s -> &_dataset_size

yeewen1018 committed:
docs(sample): document Send/Sync safety with tch-rs references

yeewen1018 committed:
refactor(dataset)!: remove old Dataset API

- Drop the old Dataset trait and its InMemoryDataset/IterableDataset implementations,
which assumed data was already in Sample form (HashMap<String, tch::Tensor>).
- This paves the way for a generic Raw->Sample pipeline where users can load
raw records (text, JSON, images, etc.) and then transform them into Samples.
- Remove dataset_bench.rs, since it needs to be rewritten against the new API.

BREAKING CHANGE: the old Dataset API has been removed.

yeewen1018 committed:
chore: remove `dataset-bench` benchmark target from Cargo.toml

yeewen1018 committed:
feat(dataset): add InMemoryDataset<Raw> and introduce Transform trait

- Implement InMemoryDataset<Raw> with new(), get(), len(), with_transform()
- Hook it up to Dataset::iter() to apply the transform pipeline
- Add unit test for InMemoryDataset

yeewen1018 committed:
feat(dataset): add DataSource trait and IterableDataset<Raw>
- Implement `IterableDataset<Raw>` to wrap one or more data sources and apply transforms
- Add unit tests verifying `IterableDataset` iteration and transform integration

yeewen1018 committed:
feat(readers/txt): implement TxtSource to stream lines from .txt files

yeewen1018 committed:
feat(readers/jsonl): implement JsonlSource to stream from .jsonl files

yeewen1018 committed:
style: cargo fmt

yeewen1018 committed:
feat(readers/image_dir): implement ImageDirSource to stream image files from a directory

yeewen1018 committed:
feat(readers/parquet): implement ParquetSource to stream RecordBatches from Parquet files

yeewen1018 committed:
feat(readers/safetensors): implement SafetensorsSource to stream u8 tensors from .safetensors files

yeewen1018 committed:
refactor(readers/parquet.rs): remove redundant code from ParquetSource

yeewen1018 committed:
ci: pull manylinux2014 image from Docker Hub for wheel build

nikg7 commented on .github/workflows/CI.yml:
```diff
@@ -47,7 +47,7 @@ jobs:
           args: --release --out dist --find-interpreter
           sccache: ${{ !startsWith(github.ref, 'refs/tags/') }}
-          manylinux: auto
+          manylinux: "manylinux2014"
```
is this change necessary? 

if yes, let's split it into a separate PR , including details why it's needed.
If not, let's revert

nikg7 commented on crates/data_preparation/src/readers/image_dir.rs:
```diff
@@ -0,0 +1,144 @@
+    /// - `dir_path`: Directory to scan (or glob pattern if needed).
+    /// - `extensions`: File extensions to include (e.g., `["jpg", "png"]`). Case-insensitive.
+    /// - `recurse`: If `true`, scans subdirectories recursively.
+    pub fn new(dir_path: impl Into<PathBuf>, extensions: &[&str], recurse: bool) -> Self {
```
nit: has the file been auto-formatted ? ("{" seems to overflow to the next line)

nikg7 commented on crates/data_preparation/src/readers/image_dir.rs:
```diff
@@ -0,0 +1,144 @@
+            Box::new(
+                WalkDir::new(&self.dir_path)
+                    .into_iter()
+                    .filter_map(|e| e.ok())
```
If `e.ok()` is `false` : What can be the reason for that  ? 

Depending on the reason we can consider whether to ignore non OK values, or fail

yeewen1018 replied:
Okay I reverted the `manylinux` pin (CI is green again), fixed the error handling in `image.rs` to propagate failures with context, and applied `cargo fmt --all`. Thank you.

yeewen1018 committed:
Revert "ci: pull manylinux2014 image from Docker Hub for wheel build"

This reverts commit 04d0617cab8e8b9c6858c38491d55967e3763bab.

yeewen1018 committed:
refactor(readers/image_dir): propagate filesystem errors with context (previously silently dropped)

yeewen1018 committed:
Merge branch 'main' into data_sources

yeewen1018 committed:
feat(readers/safetensors): add support for streaming multiple tensor dtypes

yeewen1018 committed:
refactor(readers/safetensors): optimize streaming to parse metadata once

- Replace per-tensor header re-parsing with single upfront metadata read
- Add error context for tensor conversion failures
- Add documentation for `tensor_from_view` function

yeewen1018 commented:
@nikg4  I just fixed safetensors streaming to parse metadata once (previously it was per-tensor) and added multi-dtype support. It is now ready for review - thanks!

yeewen1018 commented:
@nikg4 I just opened a PR #24 for the `Transform` trait with some basic vision/text transforms and a multimodal pipeline. It does not depend on this PR #23, but I set its base to this `data_sources` branch just to keep things in order. So no rush - you can take a look whenever you have time, thank you!
